### ISSUE TITLE
Theming: compute the primary color based on the background image

### ISSUE THREAD

**Issue by PVince81** [nextcloud/server#34560]
@skjnldsv I think you said you had an idea how to do this ?

can you post your ideas / concepts here ?

I guess the code in question (hopefully a library) will need to process all pixels of the image

**Comment by PVince81** [comment]
currently as an alternative we will let the user pick the primary color themselves: https://github.com/nextcloud/server/pull/34437

**Comment by skjnldsv** [comment]
This require an extra step:
1. Separate the config between background and color. So we can have BOTH a custom bg AND a custom colour.

Then we can address this.
I would use that in the front while uploading.
That way the user can also decide to fine-tune the colour afterwards with the color-picker
https://lokeshdhakar.com/projects/color-thief/

**Comment by jancborchardt** [comment]
> Separate the config between background and color. So we can have BOTH a custom bg AND a custom colour.

To be honest I would say we only go with autogeneration. It’s an end-user feature, it should just work, like the Material You color algo. The resulting color needs to be accessible, visually pleasing, etc – nothing that you want to necessarily fiddle around with manually.

**Comment by skjnldsv** [comment]
> To be honest I would say we only go with autogeneration. It’s an end-user feature, it should just work, like the Material You color algo. The resulting color needs to be accessible, visually pleasing, etc – nothing that you want to necessarily fiddle around with manually.


It's not an option, it's a technical requirement for this to be possible.

**Comment by skjnldsv** [comment]
@jancborchardt @nimishavijay I'm gonna use this library https://jariz.github.io/vibrant.js/
It generate a few type of colour from an image, please tell me which of the following I should use as primary:
- Vibrant
- Muted
- DarkVibrant
- DarkMuted
- LightVibrant
- LightMuted

![image](https://user-images.githubusercontent.com/14975046/196944101-71642528-b17a-4a3c-83e7-08103c119905.png)

**Comment by skjnldsv** [comment]
I guess the difference is if we want an "accent" colour or more like an "average" colour.
I would go for Vibrant (personal preference), otherwise Muted seems to be good ? :shrug:

**Comment by skjnldsv** [comment]
> Another possible idea for the future: in the color picker for theming your account, show this list of colors as "Suggested colors based on your background" in addition to the default colors.

I could do that yep :)

**Comment by jancborchardt** [comment]
Yeah, I'm also unsure between Vibrant and DarkVibrant. The others I would for sure exclude.

Vibrant seems great but not accessible as @nimishavijay said, so let's indeed go with DarkVibrant?

**Comment by jancborchardt** [comment]
> Another possible idea for the future: in the color picker for theming your account, show this list of colors as "Suggested colors based on your background" in addition to the default colors.

@nimishavijay for some reason my comment regarding that got minimized, but I said:

> To be honest I would say we only go with autogeneration. It’s an end-user feature, it should just work, like the Material You color algo. The resulting color needs to be accessible, visually pleasing, etc – nothing that you want to necessarily fiddle around with manually.

And e.g. in this case we just see that:
- DarkVibrant ticks the boxes
- Vibrant is not accessible
- The other colors are boring, or alsp not accessible

So I would say we don't need to offer them up for picking.

**Comment by skjnldsv** [comment]
Thank you both :)

**Comment by PVince81** [comment]
WIP PR here: https://github.com/nextcloud/server/pull/34696

**Comment by PVince81** [comment]
~1 week of work to get it done

**Comment by PVince81** [comment]
to clarify: 25.0.1 will already have manual selection of primary color so the base requirement was already there
so no feature regression

and this automatic detection is "nice to have"

**Comment by jancborchardt** [comment]
@skjnldsv I think for the shipped backgrounds we should continue to use the preset colors we currently have. We generated them by using the Material Design customizer, and that works pretty well (might even be the best algorithm).

**Comment by skjnldsv** [comment]
Good from me! :+1: 
Feel free to test and review https://github.com/nextcloud/server/pull/34696 then :wink:

**Comment by te-online** [comment]
Maybe this is not the right place to discuss this, feel free to point me somewhere else.

I can see how it's a nice feature to ship an accent color for the predefined background images, but when I upgraded to 25, I really felt like this was a bug.

I had selected one of the default images (the big wheel) in a previous version and suddenly my accent color went from the one that I had set globally in the theming settings to a somewhat ugly brown (I know, that's subjective 😉)

I thought, well, it's a nice feature, but I can probably change that – however, I can't seem to do that. Whenever I select this image as my background, it automatically sets the brown. Maybe I'm missing something?

It feels like a regression that we don't allow users to select their accent color separately from the background image, if we're already doing introducing so many layers to the color/background settings... 🤓
